File: backend/utils/geo.py
import math
from typing import List, Tuple

def haversine(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Distance in metres between two geographic points."""
    R = 6371000
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dp = math.radians(lat2 - lat1)
    dl = math.radians(lng2 - lng1)
    a = math.sin(dp/2)**2 + math.cos(p1)*math.cos(p2)*math.sin(dl/2)**2
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

def bearing(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Initial bearing in degrees (0-360) from point 1 to point 2."""
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dl = math.radians(lng2 - lng1)
    y = math.sin(dl) * math.cos(p2)
    x = math.cos(p1)*math.sin(p2) - math.sin(p1)*math.cos(p2)*math.cos(dl)
    return (math.degrees(math.atan2(y, x)) + 360) % 360

def heading_error(current: float, required: float) -> float:
    """Signed shortest-path heading error in degrees (-180 to +180)."""
    err = (required - current + 180) % 360 - 180
    return err

File: simulator/src/engine/geo.ts
// Ported from backend/utils/geo.py — haversine, bearing, heading_error
// Exact math preserved; projection constants copied from Python source.

const R = 6371000;

export function haversine(lat1: number, lng1: number, lat2: number, lng2: number): number {
  // Port of utils/geo.py::haversine
  const p1 = (lat1 * Math.PI) / 180;
  const p2 = (lat2 * Math.PI) / 180;
  const dp = ((lat2 - lat1) * Math.PI) / 180;
  const dl = ((lng2 - lng1) * Math.PI) / 180;
  const a = Math.sin(dp / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(Math.max(0.0, 1 - a)));
}

export function bearing(lat1: number, lng1: number, lat2: number, lng2: number): number {
  // Port of utils/geo.py::bearing
  const p1 = (lat1 * Math.PI) / 180;
  const p2 = (lat2 * Math.PI) / 180;
  const dl = ((lng2 - lng1) * Math.PI) / 180;
  const y = Math.sin(dl) * Math.cos(p2);
  const x = Math.cos(p1) * Math.sin(p2) - Math.sin(p1) * Math.cos(p2) * Math.cos(dl);
  return ((Math.atan2(y, x) * 180) / Math.PI + 360) % 360;
}

export function headingError(current: number, required: number): number {
  // Port of utils/geo.py::heading_error
  const err = ((required - current + 180) % 360) - 180;
  return err;
}

// Projection constants — copied exactly from kinematics.py / lawnmower.py
export const METERS_PER_DEG_LAT = 110540.0;
export function metersPerDegLng(lat: number): number {
  return 111320.0 * Math.cos((lat * Math.PI) / 180);
}
